### ISSUE TITLE
Channels lose the subscription when reconnects

### ISSUE THREAD

**Issue by RobertoSchneiders** [pusher-community/pusher-websocket-dotnet#8]
My code is similar to the example in this repository:
```
pusher = new Pusher("key");
pusher.Connected += pusher_Connected;
pusher.ConnectionStateChanged += _pusher_ConnectionStateChanged;
pusher.Connect();
```
```
static void pusher_Connected(object sender)
{
    Console.WriteLine("Subscribing to chat");
    chatChannel = pusher.Subscribe("chat");
    chatChannel.Subscribed += _chatChannel_Subscribed;
}
```
```
static void _chatChannel_Subscribed(object sender)
{
    Console.WriteLine("Subscribed");
    Console.WriteLine("Hi! Type 'quit' to exit, otherwise type anything to chat!");

    chatChannel.Bind("my-message", (dynamic data) =>
    {
        Console.WriteLine("[" + data.name + "] " + data.message);
    });
}
``` 

When the connection fails and PusherClient reconnects, all subscriptions are lost. 
In this particular case, this happens because of [this](https://github.com/pusher-community/pusher-websocket-dotnet/blob/master/PusherClient/Pusher.cs#L133) code on Subscribe method:
```
if (Channels.ContainsKey(channelName)){
    Trace.TraceEvent(TraceEventType.Warning, 0, "Channel '" + channelName + "' is already subscribed to. Subscription event has been ignored.");
    return Channels[channelName];
}
```

Log:
```
Connection state: Connecting
Pusher Information: 0 : Websocket opened OK.
Pusher Information: 0 : Websocket message received: {"data":"{\"activity_timeout\":120,\"socket_id\":\"2096928683.2093493135\"}","event":"pusher:conne
ction_established"}
Connection state: Connected
Subscribing to chat
Pusher Information: 0 : Sending: {"event":"pusher:subscribe","data":{"channel":"chat"}}
Pusher Information: 0 : Websocket message received: {"channel":"chat","data":{},"event":"pusher_internal:subscription_succeeded"}
Subscribed
Hi! Type 'quit' to exit, otherwise type anything to chat!
# connection failure
Connection state: Disconnected
Connection state: Connecting
Pusher Information: 0 : Websocket opened OK.
Pusher Information: 0 : Websocket message received: {"data":"{\"activity_timeout\":120,\"socket_id\":\"2470478451.2274211923\"}","event":"pusher:conne
ction_established"}
Connection state: Connected
Subscribing to chat
Pusher Warning: 0 : Channel 'chat' is already subscribed to. Subscription event has been ignored.
```

I can think in two solutions for this. 

1. Add a `Channels.Clear();` on the `Connected` event and let the responsibility to re-subscribe for those who are using the library. This will solve the problem for this code I posted, but, I'm not sure this is the right choice. 

2. Make pusher explicit re-send all channel subscriptions and update the channels. I'm not sure how to implement this.

**Comment by leggetter** [comment]
Hi Roberto,

2. is definitely the preferred option.

So you know - I don't have the bandwidth to provide fixes for this and #7 right now. But pull requests are greatly received and I'm happy to discuss possible solutions.

If you're willing to contribute then please let me know and I'll try to provide some suggestions.

**Comment by mdpye** [comment]
Just an FYI, the officially supported libraries take approach number 2 (re-subscribe automatically on reconnection), although it is not universally accepted to be the best approach (for example "spurious" subscription success events are received after reconnection).

If approach 1 is followed here, I'd suggest:
 - clearing down the client model of subscriptions on *dis*connection rather than on reconnection
 - the method of doing so is kept private; the client and server must have the same understanding of what subscriptions exist, so the clearing should be done on disconnection, and only on disconnection.

**Comment by RobertoSchneiders** [comment]
Thanks for the feedback @leggetter, @mdpye. I will try to implement this (approach 1) and send a PR. I'm not a .NET dev, so, any suggestions are welcome.

Thanks

**Comment by leggetter** [comment]
> Just an FYI, the officially supported libraries take approach number 2

@RobertoSchneiders - I don't know what I was thinking when I commented. As @mdpye pointed out, approach 2 is the one that other official libraries take. With that in mind I think it's probably the best approach as it offers a consistent experience across all libraries. Sorry for the confusion I've caused with this.

A potential way of implementing this would be to add some logic to the client to act upon connection events (other libraries tend to have a channel manager object for handling this type of thing):
https://github.com/pusher-community/pusher-websocket-dotnet/blob/master/PusherClient/Pusher.cs#L216

* Upon disconnection mark all the channels as unsubscribed. Additional logic may need to be added to stop the channels from triggering messages over the WebSocket connection (which is currently closed).
* Upon reconnections all the channels can now be subscribed to.
* If a user (application code) unsubscribes a channel whilst the connection state is disconnected then the channel should be removed and no network communication attempted.
